# **FRAMEWORK V7 DATASET MAESTRO - CATALOGO DEL CONOCIMIENTO**

## **M0. Configuración**

In [1]:
#==========================================================
# CONFIGURACIÓN
#==========================================================

import pandas as pd
import numpy as np
from pathlib import Path

#==========================================================
# DATASET OFICIAL DEL FRAMEWORK
#==========================================================

URL_DATASET = (
    "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"
)

print("Cargando Dataset Maestro V04...")

dataset_maestro_v04 = pd.read_parquet(URL_DATASET)

print("Dataset cargado correctamente.")

Cargando Dataset Maestro V04...
Dataset cargado correctamente.


## **M1 Perfil del Dataset Maestro**

In [2]:
#==========================================================
# INFORMACIÓN GENERAL DEL DATASET
#==========================================================

print("\n" + "="*90)
print("PERFIL GENERAL DEL DATASET")
print("="*90)

print(f"Registros : {dataset_maestro_v04.shape[0]:,}")
print(f"Variables : {dataset_maestro_v04.shape[1]:,}")

print("\nPrimeras variables")

print(dataset_maestro_v04.columns.tolist())


PERFIL GENERAL DEL DATASET
Registros : 528
Variables : 42

Primeras variables
['Nodo', 'Fecha', 'Precipitacion_mm', 'Temp_Max_C', 'Temp_Min_C', 'Temp_Media_C', 'Humedad_Relativa', 'Velocidad_Viento', 'Radiacion_Solar', 'Anio', 'Mes', 'ONI', 'Estado_ENSO', 'VolumenUtilDiarioMasa', 'irca', 'nivel_de_riesgo', 'POBLACION TOTAL', 'DENSIDAD POBLACIONAL', 'INDICE DE DESEMPENO INSTITUCIONAL', 'ACCESO A AGUA POTABLE ADECUADO', 'PORCENTAJE DE LA POBLACION CON ACCESO A METODOS DE SANEAMIENTO ADECUADOS', 'COBERTURA DE ACUEDUCTO URBANO', 'COBERTURA DE ACUEDUCTO RURAL', 'COBERTURA DE ALCANTARILLADO URBANO', 'COBERTURA DE ALCANTARILLADO RURAL', 'CONTINUIDAD DE ACUEDUCTO URBANO', 'CALIDAD DEL AGUA', 'AGUAS RESIDUALES TRATADAS', 'Nivel_Minimo', 'CONDUCTIVIDAD ELECTRICA', 'DEMANDA BIOQUIMICA DE OXIGENO (DBO5)', 'DEMANDA QUIMICA DE OXIGENO (DQO)', 'FOSFORO TOTAL', 'NITROGENO TOTAL', 'OXIGENO DISUELTO (OD)', 'SOLIDOS SUSPENDIDOS TOTALES', 'TEMPERATURA', 'TURBIDEZ', 'pH', 'Version_Dataset', 'IRCA_Fuente',

In [3]:
#==========================================================
# TIPOS DE VARIABLES
#==========================================================

tipos_variables = pd.DataFrame({
    "Variable": dataset_maestro_v04.columns,
    "Tipo": dataset_maestro_v04.dtypes.astype(str)
})

print("\nTipos de variables")

print(tipos_variables)


Tipos de variables
                                                                                             Variable  \
Nodo                                                                                             Nodo   
Fecha                                                                                           Fecha   
Precipitacion_mm                                                                     Precipitacion_mm   
Temp_Max_C                                                                                 Temp_Max_C   
Temp_Min_C                                                                                 Temp_Min_C   
Temp_Media_C                                                                             Temp_Media_C   
Humedad_Relativa                                                                     Humedad_Relativa   
Velocidad_Viento                                                                     Velocidad_Viento   
Radiacion_Solar                    

In [4]:
#==========================================================
# CLASIFICACIÓN DE VARIABLES
#==========================================================

variables_numericas = dataset_maestro_v04.select_dtypes(
    include=["number"]
).columns.tolist()

variables_categoricas = dataset_maestro_v04.select_dtypes(
    include=["object","category"]
).columns.tolist()

variables_fecha = dataset_maestro_v04.select_dtypes(
    include=["datetime64[ns]"]
).columns.tolist()

In [5]:
print("\nResumen")

print(f"Variables numéricas    : {len(variables_numericas)}")
print(f"Variables categóricas  : {len(variables_categoricas)}")
print(f"Variables fecha        : {len(variables_fecha)}")


Resumen
Variables numéricas    : 35
Variables categóricas  : 6
Variables fecha        : 1


In [6]:
#==========================================================
# MEMORIA
#==========================================================

memoria = dataset_maestro_v04.memory_usage(deep=True).sum()/1024**2

print(f"\nMemoria utilizada : {memoria:.2f} MB")


Memoria utilizada : 0.32 MB


In [7]:
#==========================================================
# DUPLICADOS
#==========================================================

duplicados = dataset_maestro_v04.duplicated().sum()

print(f"\nRegistros duplicados : {duplicados}")


Registros duplicados : 0


In [8]:
#==========================================================
# COMPLETITUD GENERAL
#==========================================================

total_celdas = (
    dataset_maestro_v04.shape[0] *
    dataset_maestro_v04.shape[1]
)

celdas_validas = dataset_maestro_v04.notna().sum().sum()

completitud = (
    celdas_validas /
    total_celdas
)*100

print(f"\nCompletitud general : {completitud:.2f}%")


Completitud general : 68.16%


In [9]:
#==========================================================
# RESUMEN GENERAL
#==========================================================

perfil_general = pd.DataFrame({

    "Indicador":[
        "Registros",
        "Variables",
        "Variables numéricas",
        "Variables categóricas",
        "Variables fecha",
        "Duplicados",
        "Memoria (MB)",
        "Completitud (%)"
    ],

    "Valor":[
        dataset_maestro_v04.shape[0],
        dataset_maestro_v04.shape[1],
        len(variables_numericas),
        len(variables_categoricas),
        len(variables_fecha),
        duplicados,
        round(memoria,2),
        round(completitud,2)
    ]

})

print(perfil_general)

               Indicador   Valor
0              Registros  528.00
1              Variables   42.00
2    Variables numéricas   35.00
3  Variables categóricas    6.00
4        Variables fecha    1.00
5             Duplicados    0.00
6           Memoria (MB)    0.32
7        Completitud (%)   68.16


In [10]:
#==========================================================
# EXPORTAR RESULTADOS
#==========================================================
#
# Este bloque genera los productos del Script_Perfil_General_Dataset.
#
# perfil_general_dataset.csv
#     Tabla resumen con los principales indicadores del
#     Dataset Maestro V04.
#
# perfil_general_dataset.xlsx
#     Versión en Excel para análisis y documentación.
#
# perfil_general_dataset.md
#     Resumen en formato Markdown para incorporar
#     directamente en la tesis o documentación técnica.
#
#==========================================================

perfil_general.to_csv(
    "Perfil_general_Dataset_Maestro_Framework_v04.csv",
    index=False,
    encoding="utf-8-sig"
)

perfil_general.to_excel(
    "Perfil_general_Dataset_Maestro_Framework_v04.xlsx",
    index=False
)

perfil_general.to_markdown(
    "Perfil_general_Dataset_Maestro_Framework_v04.md",
    index=False
)

print("\nArchivos generados correctamente:")
print("   Perfil_general_Dataset_Maestro_Framework_v04.csv")
print("   Perfil_general_Dataset_Maestro_Framework_v04.xlsx")
print("   Perfil_general_Dataset_Maestro_Framework_v04.md")


Archivos generados correctamente:
   Perfil_general_Dataset_Maestro_Framework_v04.csv
   Perfil_general_Dataset_Maestro_Framework_v04.xlsx
   Perfil_general_Dataset_Maestro_Framework_v04.md


## **M2. Calidad de Datos del Dataset Maestro V04**

In [11]:
#==========================================================
# SCRIPT
# CALIDAD DE DATOS
#==========================================================

import pandas as pd
import numpy as np

URL_DATASET = (
    "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"
)

print("Cargando Dataset Maestro V04...")

dataset_maestro_v04 = pd.read_parquet(URL_DATASET)

print("Dataset cargado correctamente.")

Cargando Dataset Maestro V04...
Dataset cargado correctamente.


In [12]:
#==========================================================
# PERFIL DE VARIABLES
#==========================================================

reporte = []

In [13]:
for variable in dataset_maestro_v04.columns:

    serie = dataset_maestro_v04[variable]

    total = len(serie)

    nulos = serie.isna().sum()

    completos = total - nulos

    cobertura = completos / total * 100

    unicos = serie.nunique(dropna=True)

    duplicados = completos - unicos

    tipo = str(serie.dtype)

    porcentaje_nulos = nulos / total * 100

    porcentaje_ceros = (
        (serie == 0).sum() / total * 100
        if pd.api.types.is_numeric_dtype(serie)
        else np.nan
    )

    reporte.append({
        "Variable":variable,
        "Tipo":tipo,
        "Registros":total,
        "Completos":completos,
        "Nulos":nulos,
        "% Cobertura":round(cobertura,2),
        "% Nulos":round(porcentaje_nulos,2),
        "Valores Únicos":unicos,
        "Duplicados":duplicados,
        "% Ceros":round(porcentaje_ceros,2)
    })

In [14]:
calidad = pd.DataFrame(reporte)

In [15]:
def clasificacion(cobertura):

    if cobertura >= 95:
        return "A - Excelente"

    elif cobertura >= 80:
        return "B - Buena"

    elif cobertura >= 50:
        return "C - Media"

    else:
        return "D - Baja"

In [16]:
calidad["Calificación"] = calidad["% Cobertura"].apply(clasificacion)

In [17]:
def recomendacion(fila):

    cobertura = fila["% Cobertura"]

    if cobertura >= 95:
        return "Lista para modelado"

    elif cobertura >= 80:
        return "Imputación simple"

    elif cobertura >= 50:
        return "Imputación avanzada"

    else:
        return "No recomendable para entrenamiento"

In [18]:
calidad["Recomendación"] = calidad.apply(
    recomendacion,
    axis=1
)

In [19]:
calidad = calidad.sort_values(
    "% Cobertura",
    ascending=False
)

In [20]:
print("\n")

print("="*110)

print("CALIDAD GENERAL DEL DATASET")

print("="*110)

print(calidad)



CALIDAD GENERAL DEL DATASET
                                             Variable            Tipo  \
0                                                Nodo          object   
1                                               Fecha  datetime64[ns]   
2                                    Precipitacion_mm         float64   
3                                          Temp_Max_C         float64   
4                                          Temp_Min_C         float64   
5                                        Temp_Media_C         float64   
6                                    Humedad_Relativa         float64   
7                                    Velocidad_Viento         float64   
8                                     Radiacion_Solar         float64   
9                                                Anio           int32   
10                                                Mes           int32   
11                                                ONI         float64   
12                   

In [21]:
print("\n")

print("="*70)

print("RESUMEN")

print("="*70)

print(calidad["Calificación"].value_counts())



RESUMEN
Calificación
A - Excelente    27
D - Baja         15
Name: count, dtype: int64


In [22]:
#==========================================================
# EXPORTAR RESULTADOS DEL PERFIL DE CALIDAD
#==========================================================
#
# Módulo 5 – Análisis Exploratorio de Datos (EDA)
# Script: Calidad de Datos del Dataset Maestro V04
#
# Descripción:
#
#   Se exporta el perfil de calidad de cada variable del
#   Dataset Maestro V04.
#
#   El perfil incluye información como:
#
#       • Nombre de la variable
#       • Número de registros disponibles
#       • Valores faltantes
#       • Porcentaje de cobertura
#       • Clasificación de calidad
#
# Archivos generados:
#
#   calidad_variables.csv
#       Archivo interoperable para análisis y trazabilidad.
#
#   calidad_variables.xlsx
#       Archivo para revisión manual y documentación.
#
#   calidad_variables.md
#       Archivo en formato Markdown para incorporación
#       directa en la tesis y documentación técnica.
#
#==========================================================

calidad.to_csv(
    "Calidad_Variables_Dataset_Maestro_Framework_v04.csv",
    index=False,
    encoding="utf-8-sig"
)

calidad.to_excel(
    "Calidad_Variables_Dataset_Maestro_Framework_v04.xlsx",
    index=False
)

calidad.to_markdown(
    "Calidad_Variables_Dataset_Maestro_Framework_v04.md",
    index=False
)

print("\nArchivos generados correctamente:")
print("   Calidad_Variables_Dataset_Maestro_Framework_v04.csv")
print("   Calidad_Variables_Dataset_Maestro_Framework_v04.xlsx")
print("   Calidad_Variables_Dataset_Maestro_Framework_v04.md")


Archivos generados correctamente:
   Calidad_Variables_Dataset_Maestro_Framework_v04.csv
   Calidad_Variables_Dataset_Maestro_Framework_v04.xlsx
   Calidad_Variables_Dataset_Maestro_Framework_v04.md


## **M3. Estadisticas Descriptivas del Dataset Maestro V04**

In [23]:
import pandas as pd
import numpy as np

URL_DATASET = (
    "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"
)

print("="*70)
print("ESTADÍSTICAS DESCRIPTIVAS")
print("="*70)

dataset_maestro_v04 = pd.read_parquet(URL_DATASET)

#----------------------------------------------------------
# Seleccionar únicamente variables numéricas
#----------------------------------------------------------

variables_numericas = dataset_maestro_v04.select_dtypes(include=np.number)

print("\nVariables numéricas encontradas:", variables_numericas.shape[1])

#----------------------------------------------------------
# Estadísticas
#----------------------------------------------------------

estadisticas = pd.DataFrame()

estadisticas["Tipo"] = variables_numericas.dtypes

estadisticas["Registros"] = variables_numericas.count()

estadisticas["Faltantes"] = variables_numericas.isna().sum()

estadisticas["Cobertura (%)"] = (
    estadisticas["Registros"] /
    len(dataset_maestro_v04) * 100
).round(2)

estadisticas["Media"] = variables_numericas.mean()

estadisticas["Mediana"] = variables_numericas.median()

estadisticas["Desv.Std"] = variables_numericas.std()

estadisticas["Varianza"] = variables_numericas.var()

estadisticas["Mínimo"] = variables_numericas.min()

estadisticas["P25"] = variables_numericas.quantile(.25)

estadisticas["P75"] = variables_numericas.quantile(.75)

estadisticas["Máximo"] = variables_numericas.max()

estadisticas["Valores únicos"] = variables_numericas.nunique()

estadisticas["Coeficiente Variación (%)"] = (
    estadisticas["Desv.Std"] /
    estadisticas["Media"] * 100
).round(2)

print()
print(estadisticas)

ESTADÍSTICAS DESCRIPTIVAS

Variables numéricas encontradas: 35

                                                       Tipo  Registros  \
Precipitacion_mm                                    float64        528   
Temp_Max_C                                          float64        528   
Temp_Min_C                                          float64        528   
Temp_Media_C                                        float64        528   
Humedad_Relativa                                    float64        528   
Velocidad_Viento                                    float64        528   
Radiacion_Solar                                     float64        528   
Anio                                                  int32        528   
Mes                                                   int32        528   
ONI                                                 float64        528   
VolumenUtilDiarioMasa                                 int64        528   
irca                                            

In [24]:
#==========================================================
# EXPORTAR ESTADÍSTICAS DESCRIPTIVAS
#==========================================================
#
# Se exporta el resumen estadístico de todas las variables
# numéricas del Dataset Maestro V04.
#
# Contenido:
#     • Tendencia central
#     • Dispersión
#     • Cobertura
#     • Valores únicos
#     • Variabilidad
#
#==========================================================

estadisticas.to_csv(
    "Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.csv",
    index=True,
    encoding="utf-8-sig"
)

estadisticas.to_excel(
    "Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.xlsx"
)

estadisticas.to_markdown(
    "Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.md"
)

print("\nArchivos generados correctamente:")
print("   Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.csv")
print("   Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.xlsx")
print("   Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.md")


Archivos generados correctamente:
   Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.csv
   Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.xlsx
   Estadisticas_Descriptivas_Dataset_Maestro_Framework_v04.md


## **M4. Distribuciones Univariadas del Dataset Maestro V04**

In [25]:
#==========================================================
# FRAMEWORK PARA LA GESTIÓN INTELIGENTE DEL RECURSO HÍDRICO
#
# Análisis Exploratorio de Datos (EDA)
#
# Script
# Distribuciones Univariadas
#
# Objetivo
# ---------
# Analizar la distribución estadística de todas las
# variables numéricas del Dataset Maestro V04 mediante
# histogramas.
#
# Para cada variable se genera:
#
#   • Histograma
#   • Línea de la media
#   • Línea de la mediana
#   • Estadísticos básicos
#
# Los gráficos serán utilizados en la tesis y para
# apoyar la selección de variables del Framework.
#
# Fuente:
# Dataset Maestro V04 (GitHub)
#
#==========================================================

import os

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

#==========================================================
# CARGAR DATASET MAESTRO V04
#==========================================================

print("Cargando Dataset Maestro V04...")

url = (
    "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"
)

dataset = pd.read_parquet(url)

print("Dataset cargado correctamente.")

Cargando Dataset Maestro V04...
Dataset cargado correctamente.


In [26]:
#==========================================================
# CREAR CARPETA DE RESULTADOS
#==========================================================

carpeta = "EDA_Distribuciones"

os.makedirs(carpeta, exist_ok=True)

In [27]:
variables = dataset.select_dtypes(include=np.number).columns

print()

print("Variables numéricas encontradas:", len(variables))


Variables numéricas encontradas: 35


In [28]:
#==========================================================
# HISTOGRAMAS
#==========================================================

for variable in variables:

    datos = dataset[variable].dropna()

    if len(datos) == 0:
        continue

    plt.figure(figsize=(10,6))

    plt.hist(
        datos,
        bins=20,
        edgecolor="black"
    )

    media = datos.mean()
    mediana = datos.median()

    plt.axvline(
        media,
        color="red",
        linestyle="--",
        linewidth=2,
        label=f"Media = {media:.2f}"
    )

    plt.axvline(
        mediana,
        color="green",
        linestyle=":",
        linewidth=2,
        label=f"Mediana = {mediana:.2f}"
    )

    plt.title(variable)

    plt.xlabel(variable)

    plt.ylabel("Frecuencia")

    plt.legend()

    plt.grid(alpha=0.30)

    nombre = variable

    for c in [
        "/","\\",":","*","?","<",">","|","(",")"
    ]:
        nombre = nombre.replace(c,"")

    plt.savefig(
        f"{carpeta}/{nombre}.png",
        dpi=300,
        bbox_inches="tight"
    )

    plt.close()

print()

print("Histogramas generados correctamente.")


Histogramas generados correctamente.


In [29]:
#==========================================================
# EXPORTAR RESUMEN
#==========================================================

resumen = pd.DataFrame({

    "Variable": variables,

    "Registros":
        [dataset[v].count() for v in variables],

    "Media":
        [dataset[v].mean() for v in variables],

    "Mediana":
        [dataset[v].median() for v in variables],

    "DesvStd":
        [dataset[v].std() for v in variables]

})

resumen.to_csv(
    "EDA_Distribuciones/Resumen_Distribuciones.csv",
    index=False,
    encoding="utf-8-sig"
)

print()

print("Archivo generado:")

print("   Resumen_Distribuciones.csv")


Archivo generado:
   Resumen_Distribuciones.csv


## **M5. Distribuciones Univariadas del Dataset Maestro V04 - Fichas Estadísticas Automáticas**

In [30]:
resultado_univariado = []

In [31]:
import os
from pathlib import Path

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

#==========================================================
# DATASET MAESTRO V04
#==========================================================

url = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"

print("Cargando Dataset Maestro V04...")

dataset = pd.read_parquet(url)

print("Dataset cargado correctamente.")

print(dataset.shape)

Cargando Dataset Maestro V04...
Dataset cargado correctamente.
(528, 42)


In [32]:
#==========================================================
# CARPETAS DE RESULTADOS
#==========================================================

CARPETA = Path("Resultados_EDA")

CARPETA.mkdir(exist_ok=True)

(Path("Resultados_EDA/Fichas")).mkdir(exist_ok=True)

(Path("Resultados_EDA/Histogramas")).mkdir(exist_ok=True)

(Path("Resultados_EDA/Boxplots")).mkdir(exist_ok=True)

print("Carpetas creadas.")

Carpetas creadas.


In [33]:
#==========================================================
# VARIABLES NUMÉRICAS
#==========================================================

variables = dataset.select_dtypes(include=np.number).columns.tolist()

print()

print("Variables encontradas:",len(variables))


Variables encontradas: 35


In [34]:
#==========================================================
# DICCIONARIO OFICIAL DE VARIABLES
# FRAMEWORK GESTIÓN INTELIGENTE DEL RECURSO HÍDRICO
#==========================================================

diccionario_variables = {

    #======================================================
    # IDENTIFICACIÓN
    #======================================================

    "Anio": {
        "Origen":"Framework",
        "Categoria":"Temporal",
        "Unidad":"Año"
    },

    "Mes": {
        "Origen":"Framework",
        "Categoria":"Temporal",
        "Unidad":"Mes"
    },

    #======================================================
    # CLIMATOLOGÍA
    #======================================================

    "Precipitacion_mm":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"mm"
    },

    "Temp_Media_C":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"°C"
    },

    "Temp_Min_C":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"°C"
    },

    "Temp_Max_C":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"°C"
    },

    "Humedad_Relativa":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"%"
    },

    "Radiacion_Solar":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"W/m²"
    },

    "Velocidad_Viento":{
        "Origen":"IDEAM",
        "Categoria":"Climatología",
        "Unidad":"m/s"
    },

    #======================================================
    # ENSO
    #======================================================

    "ONI":{
        "Origen":"NOAA",
        "Categoria":"ENSO",
        "Unidad":"Índice"
    },

    #======================================================
    # DEMOGRAFÍA
    #======================================================

    "POBLACION TOTAL":{
        "Origen":"DANE",
        "Categoria":"Demografía",
        "Unidad":"Habitantes"
    },

    "DENSIDAD POBLACIONAL":{
        "Origen":"DANE",
        "Categoria":"Demografía",
        "Unidad":"Hab/km²"
    },

    #======================================================
    # RECURSO HÍDRICO
    #======================================================

    "Nivel_Minimo":{
        "Origen":"CAR",
        "Categoria":"Hidrología",
        "Unidad":"m"
    },

    "VolumenUtilDiarioMasa":{
        "Origen":"CAR",
        "Categoria":"Hidrología",
        "Unidad":"m³"
    },

    #======================================================
    # CALIDAD DEL AGUA
    #======================================================

    "DEMANDA BIOQUIMICA DE OXIGENO (DBO5)":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"mg/L"
    },

    "DEMANDA QUIMICA DE OXIGENO (DQO)":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"mg/L"
    },

    "FOSFORO TOTAL":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"mg/L"
    },

    "NITROGENO TOTAL":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"mg/L"
    },

    "OXIGENO DISUELTO (OD)":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"mg/L"
    },

    "CONDUCTIVIDAD ELECTRICA":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"µS/cm"
    },

    "SOLIDOS SUSPENDIDOS TOTALES":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"mg/L"
    },

    "TURBIDEZ":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"NTU"
    },

    "TEMPERATURA":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"°C"
    },

    "pH":{
        "Origen":"CAR",
        "Categoria":"Calidad del Agua",
        "Unidad":"pH"
    },

    #======================================================
    # AGUA POTABLE
    #======================================================

    "irca":{
        "Origen":"SIVICAP",
        "Categoria":"Agua Potable",
        "Unidad":"%"
    },

    "CALIDAD DEL AGUA":{
        "Origen":"TerriData",
        "Categoria":"Agua Potable",
        "Unidad":"Índice"
    },

    #======================================================
    # INFRAESTRUCTURA
    #======================================================

    "CONTINUIDAD DE ACUEDUCTO URBANO":{
        "Origen":"TerriData",
        "Categoria":"Infraestructura",
        "Unidad":"Horas/día"
    },

    "COBERTURA DE ACUEDUCTO URBANO":{
        "Origen":"TerriData",
        "Categoria":"Infraestructura",
        "Unidad":"%"
    },

    "COBERTURA DE ACUEDUCTO RURAL":{
        "Origen":"TerriData",
        "Categoria":"Infraestructura",
        "Unidad":"%"
    },

    "COBERTURA DE ALCANTARILLADO URBANO":{
        "Origen":"TerriData",
        "Categoria":"Infraestructura",
        "Unidad":"%"
    },

    "COBERTURA DE ALCANTARILLADO RURAL":{
        "Origen":"TerriData",
        "Categoria":"Infraestructura",
        "Unidad":"%"
    },

    "AGUAS RESIDUALES TRATADAS":{
        "Origen":"TerriData",
        "Categoria":"Saneamiento",
        "Unidad":"%"
    },

    #======================================================
    # GOBERNANZA
    #======================================================

    "INDICE DE DESEMPENO INSTITUCIONAL":{
        "Origen":"DNP",
        "Categoria":"Gobernanza",
        "Unidad":"Índice"
    }

}

In [35]:
#==========================================================
# FUNCIÓN
# OBTENER METADATOS DE UNA VARIABLE
#==========================================================

def obtener_metadatos(variable):

    if variable in diccionario_variables:

        return diccionario_variables[variable]

    return {

        "Origen":"No definido",

        "Categoria":"No definida",

        "Unidad":"-"

    }

In [36]:
#==========================================================
# FUNCIÓN
# CALCULAR ESTADÍSTICOS DE UNA VARIABLE
#==========================================================

def calcular_estadisticos(serie, total_registros):

    #---------------------------------------------
    # Eliminar valores faltantes
    #---------------------------------------------

    serie = serie.dropna()

    observaciones = len(serie)

    faltantes = total_registros - observaciones

    cobertura = (
        observaciones /
        total_registros * 100
    )

    #---------------------------------------------
    # Estadísticos descriptivos
    #---------------------------------------------

    media = serie.mean()

    mediana = serie.median()

    moda = (
        serie.mode().iloc[0]
        if not serie.mode().empty
        else np.nan
    )

    desviacion = serie.std()

    varianza = serie.var()

    minimo = serie.min()

    maximo = serie.max()

    rango = maximo - minimo

    #---------------------------------------------
    # Percentiles
    #---------------------------------------------

    p5 = serie.quantile(0.05)

    q1 = serie.quantile(0.25)

    q2 = serie.quantile(0.50)

    q3 = serie.quantile(0.75)

    p95 = serie.quantile(0.95)

    iqr = q3 - q1

    #---------------------------------------------
    # Coeficiente de variación
    #---------------------------------------------

    if media != 0:

        coeficiente_variacion = (
            desviacion /
            media
        ) * 100

    else:

        coeficiente_variacion = np.nan

    #---------------------------------------------
    # Asimetría y Curtosis
    #---------------------------------------------

    asimetria = serie.skew()

    curtosis = serie.kurt()

    #---------------------------------------------
    # Detección de Outliers (IQR)
    #---------------------------------------------

    limite_inferior = q1 - 1.5 * iqr

    limite_superior = q3 + 1.5 * iqr

    outliers = serie[
        (serie < limite_inferior) |
        (serie > limite_superior)
    ]

    cantidad_outliers = len(outliers)

    porcentaje_outliers = (
        cantidad_outliers /
        observaciones * 100
        if observaciones > 0
        else 0
    )

    #---------------------------------------------
    # Retornar resultados
    #---------------------------------------------

    return {

        "Observaciones": observaciones,

        "Faltantes": faltantes,

        "Cobertura": cobertura,

        "Media": media,

        "Mediana": mediana,

        "Moda": moda,

        "Desviacion": desviacion,

        "Varianza": varianza,

        "Minimo": minimo,

        "Maximo": maximo,

        "Rango": rango,

        "P5": p5,

        "Q1": q1,

        "Q2": q2,

        "Q3": q3,

        "P95": p95,

        "IQR": iqr,

        "CoeficienteVariacion": coeficiente_variacion,

        "Asimetria": asimetria,

        "Curtosis": curtosis,

        "Outliers": cantidad_outliers,

        "PorcentajeOutliers": porcentaje_outliers

    }

In [37]:
#==========================================================
# FUNCIÓN
# INTERPRETACIÓN AUTOMÁTICA DE LA VARIABLE
#==========================================================

def interpretar_variable(est):

    interpretacion = []

    #------------------------------------------------------
    # Cobertura
    #------------------------------------------------------

    cobertura = est["Cobertura"]

    if cobertura >= 95:
        interpretacion.append(
            "La variable presenta una cobertura excelente (>95%)."
        )

    elif cobertura >= 80:
        interpretacion.append(
            "La cobertura de la variable es adecuada para análisis estadísticos."
        )

    elif cobertura >= 50:
        interpretacion.append(
            "La cobertura es moderada; los resultados deben interpretarse con precaución."
        )

    else:
        interpretacion.append(
            "La variable presenta una cobertura baja, por lo que los resultados son limitados."
        )

    #------------------------------------------------------
    # Asimetría
    #------------------------------------------------------

    skew = est["Asimetria"]

    if abs(skew) < 0.5:

        interpretacion.append(
            "La distribución es aproximadamente simétrica."
        )

    elif abs(skew) < 1:

        if skew > 0:

            interpretacion.append(
                "La distribución presenta un sesgo moderado hacia la derecha."
            )

        else:

            interpretacion.append(
                "La distribución presenta un sesgo moderado hacia la izquierda."
            )

    else:

        if skew > 0:

            interpretacion.append(
                "La distribución presenta un fuerte sesgo hacia la derecha."
            )

        else:

            interpretacion.append(
                "La distribución presenta un fuerte sesgo hacia la izquierda."
            )

    #------------------------------------------------------
    # Curtosis
    #------------------------------------------------------

    kurt = est["Curtosis"]

    if kurt > 3:

        interpretacion.append(
            "La distribución presenta colas pesadas (leptocúrtica), con mayor probabilidad de valores extremos."
        )

    elif kurt < 0:

        interpretacion.append(
            "La distribución es más plana que una distribución normal (platicúrtica)."
        )

    else:

        interpretacion.append(
            "La curtosis es compatible con una distribución de comportamiento intermedio."
        )

    #------------------------------------------------------
    # Variabilidad
    #------------------------------------------------------

    cv = est["CoeficienteVariacion"]

    if np.isnan(cv):

        interpretacion.append(
            "No fue posible calcular el coeficiente de variación."
        )

    elif cv < 20:

        interpretacion.append(
            "La variabilidad de la variable es baja."
        )

    elif cv < 50:

        interpretacion.append(
            "La variabilidad observada es moderada."
        )

    else:

        interpretacion.append(
            "La variable presenta una alta variabilidad."
        )

    #------------------------------------------------------
    # Valores atípicos
    #------------------------------------------------------

    out = est["PorcentajeOutliers"]

    if out == 0:

        interpretacion.append(
            "No se detectaron valores atípicos."
        )

    elif out < 5:

        interpretacion.append(
            "Se identificó una baja proporción de valores atípicos."
        )

    elif out < 10:

        interpretacion.append(
            "La variable presenta una cantidad moderada de valores atípicos."
        )

    else:

        interpretacion.append(
            "Existe una presencia importante de valores atípicos."
        )

    #------------------------------------------------------
    # Conclusión general
    #------------------------------------------------------

    if cobertura < 50:

        interpretacion.append(
            "Debido a su baja cobertura, esta variable no es recomendable para procesos de modelado sin aplicar estrategias adicionales de tratamiento."
        )

    return "\n\n".join(interpretacion)

In [38]:
#==========================================================
# FUNCIÓN
# GENERAR FICHA ESTADÍSTICA
#==========================================================

def generar_ficha(variable, serie):

    #---------------------------------------------
    # Eliminar valores faltantes
    #---------------------------------------------

    serie = serie.dropna()

    if len(serie) == 0:
        return

    #---------------------------------------------
    # Estadísticos
    #---------------------------------------------

    est = calcular_estadisticos(serie, len(dataset))

    texto = interpretar_variable(est)

    meta = obtener_metadatos(variable)

    #---------------------------------------------
    # Crear figura
    #---------------------------------------------

    fig = plt.figure(
        figsize=(16,12)
    )

    gs = fig.add_gridspec(
        4,
        2,
        height_ratios=[0.45,2.4,0.8,1.3],
        width_ratios=[1,1]
    )

    #---------------------------------------------
    # Encabezado
    #---------------------------------------------

    ax_titulo = fig.add_subplot(gs[0,:])

    ax_titulo.axis("off")

    titulo = (
        "FRAMEWORK PARA LA GESTIÓN INTELIGENTE "
        "DEL RECURSO HÍDRICO\n"
        "Ficha Estadística de Variable"
    )

    ax_titulo.text(
        0.5,
        0.70,
        titulo,
        fontsize=18,
        fontweight="bold",
        ha="center"
    )

    informacion = (
        f"Variable: {variable}\n"
        f"Origen: {meta['Origen']}\n"
        f"Categoría: {meta['Categoria']}\n"
        f"Unidad: {meta['Unidad']}\n"
        f"Cobertura: {est['Cobertura']:.2f}%"
    )

    ax_titulo.text(
        0.02,
        0.05,
        informacion,
        fontsize=11,
        ha="left"
    )

    #======================================================
    # HISTOGRAMA
    #======================================================

    ax_hist = fig.add_subplot(gs[1,:])

    ax_hist.hist(
        serie,
        bins="auto",
        edgecolor="black",
        alpha=0.75
    )

    # Línea de la media
    ax_hist.axvline(
        est["Media"],
        color="red",
        linestyle="--",
        linewidth=2,
        label=f"Media = {est['Media']:.2f}"
    )

    # Línea de la mediana
    ax_hist.axvline(
        est["Mediana"],
        color="green",
        linestyle="-.",
        linewidth=2,
        label=f"Mediana = {est['Mediana']:.2f}"
    )

    ax_hist.set_title(
        "Distribución de la Variable",
        fontsize=13,
        fontweight="bold"
    )

    ax_hist.set_xlabel(variable)

    ax_hist.set_ylabel("Frecuencia")

    ax_hist.grid(
        alpha=0.30
    )

    ax_hist.legend()

    #======================================================
    # BOXPLOT
    #======================================================

    ax_box = fig.add_subplot(gs[2,:])

    ax_box.boxplot(
        serie,
        vert=False,
        patch_artist=True,
        widths=0.50,
        boxprops=dict(
            facecolor="lightblue",
            color="navy"
        ),
        medianprops=dict(
            color="red",
            linewidth=2
        ),
        whiskerprops=dict(
            color="black"
        ),
        capprops=dict(
            color="black"
        ),
        flierprops=dict(
            marker='o',
            markerfacecolor='orange',
            markeredgecolor='black',
            markersize=5
        )
    )

    ax_box.set_title(
        "Boxplot",
        fontsize=13,
        fontweight="bold"
    )

    ax_box.set_xlabel(variable)

    ax_box.grid(alpha=0.30)

    #======================================================
    # TABLA ESTADÍSTICA
    #======================================================

    ax_tabla = fig.add_subplot(gs[3,0])

    ax_tabla.axis("off")

    tabla = [

        ["Observaciones", est["Observaciones"]],

        ["Cobertura (%)", f'{est["Cobertura"]:.2f}'],

        ["Media", f'{est["Media"]:.3f}'],

        ["Mediana", f'{est["Mediana"]:.3f}'],

        ["Desv. Estándar", f'{est["Desviacion"]:.3f}'],

        ["Varianza", f'{est["Varianza"]:.3f}'],

        ["Coef. Variación (%)", f'{est["CoeficienteVariacion"]:.2f}'],

        ["Mínimo", f'{est["Minimo"]:.3f}'],

        ["Q1", f'{est["Q1"]:.3f}'],

        ["Q3", f'{est["Q3"]:.3f}'],

        ["IQR", f'{est["IQR"]:.3f}'],

        ["Máximo", f'{est["Maximo"]:.3f}'],

        ["Asimetría", f'{est["Asimetria"]:.3f}'],

        ["Curtosis", f'{est["Curtosis"]:.3f}'],

        ["Outliers", est["Outliers"]],

        ["% Outliers", f'{est["PorcentajeOutliers"]:.2f}']

    ]

    tabla_est = ax_tabla.table(

        cellText=tabla,

        colLabels=["Indicador","Valor"],

        cellLoc="left",

        loc="center"

    )

    tabla_est.auto_set_font_size(False)

    tabla_est.set_fontsize(9)

    tabla_est.scale(1.25,1.60)

    #======================================================
    # INTERPRETACIÓN AUTOMÁTICA
    #======================================================

    ax_texto = fig.add_subplot(gs[3,1])

    ax_texto.axis("off")

    interpretacion = []

    #--------------------------------------------------
    # Cobertura
    #--------------------------------------------------

    if est["Cobertura"] >= 95:
        interpretacion.append(
            "• Cobertura excelente."
        )

    elif est["Cobertura"] >= 80:
        interpretacion.append(
            "• Cobertura alta."
        )

    elif est["Cobertura"] >= 50:
        interpretacion.append(
            "• Cobertura moderada."
        )

    else:
        interpretacion.append(
            "• Cobertura baja."
        )

    #--------------------------------------------------
    # Asimetría
    #--------------------------------------------------

    if abs(est["Asimetria"]) < 0.5:

        interpretacion.append(
            "• Distribución aproximadamente simétrica."
        )

    elif est["Asimetria"] > 0.5:

        interpretacion.append(
            "• Distribución sesgada hacia la derecha."
        )

    else:

        interpretacion.append(
            "• Distribución sesgada hacia la izquierda."
        )

    #--------------------------------------------------
    # Curtosis
    #--------------------------------------------------

    if est["Curtosis"] > 3:

        interpretacion.append(
            "• Presenta colas pesadas y posibles valores extremos."
        )

    elif est["Curtosis"] < 0:

        interpretacion.append(
            "• Distribución relativamente plana."
        )

    else:

        interpretacion.append(
            "• Curtosis cercana a una distribución normal."
        )

    #--------------------------------------------------
    # Variabilidad
    #--------------------------------------------------

    if est["CoeficienteVariacion"] < 20:

        interpretacion.append(
            "• Variabilidad baja."
        )

    elif est["CoeficienteVariacion"] < 50:

        interpretacion.append(
            "• Variabilidad moderada."
        )

    else:

        interpretacion.append(
            "• Alta variabilidad."
        )

    #--------------------------------------------------
    # Outliers
    #--------------------------------------------------

    if est["PorcentajeOutliers"] < 1:

        interpretacion.append(
            "• No se detectan valores atípicos relevantes."
        )

    elif est["PorcentajeOutliers"] < 5:

        interpretacion.append(
            "• Se detectan algunos valores atípicos."
        )

    else:

        interpretacion.append(
            "• Existe una presencia importante de valores atípicos."
        )

    #--------------------------------------------------
    # Texto
    #--------------------------------------------------

    ax_texto.text(

        0,

        1,

        "\n\n".join(interpretacion),

        fontsize=10,

        verticalalignment="top",

        wrap=True

    )

    ax_texto.set_title(

        "Interpretación Automática",

        fontsize=12,

        fontweight="bold"

    )

    #==========================================================
    # PIE DE PÁGINA
    #==========================================================

    fig.text(

        0.50,

        0.02,

        "Framework para la Gestión Inteligente del Recurso Hídrico\n"
        "Dataset Maestro V04\n"
        "Maestría en Ciencia de Datos\n"
        "Juan Carlos Riatiga Peña",

        ha="center",

        fontsize=9,

        color="gray"

    )


    #==========================================================
    # AJUSTAR ESPACIOS
    #==========================================================

    plt.tight_layout(
        rect=[0,0.04,1,0.97]
    )


    #==========================================================
    # EXPORTAR FICHA
    #==========================================================

    nombre_archivo = (

        variable
        .replace("/","-")
        .replace("\\","-")
        .replace(" ","_")
        .replace("(","")
        .replace(")","")

    )

    # Define carpeta_resultados before using it
    carpeta_resultados = CARPETA / "Fichas"

    ruta = carpeta_resultados / f"{nombre_archivo}.png"

    plt.savefig(

        ruta,

        dpi=300,

        bbox_inches="tight"

    )

    plt.close(fig)

    print(f"Ficha generada -> {ruta.name}")

In [39]:
resultado_univariado = []

print()

print("="*80)
print("INICIANDO ANÁLISIS UNIVARIADO")
print("="*80)

for variable in variables:

    print()

    print("-"*70)
    print(variable)
    print("-"*70)

    serie = dataset[variable]

    #----------------------------------------------
    # Calcular estadísticas
    #----------------------------------------------

    est = calcular_estadisticos(serie, len(dataset))

    meta = obtener_metadatos(variable)

    #----------------------------------------------
    # Generar ficha estadística
    #----------------------------------------------

    generar_ficha(
        variable,
        serie
    )

    #----------------------------------------------
    # Agregar al resumen
    #----------------------------------------------

    resultado_univariado.append({

        "Variable":variable,

        "Origen":meta["Origen"],

        "Categoria":meta["Categoria"],

        "Unidad":meta["Unidad"],

        "Observaciones":est["Observaciones"],

        "Cobertura (%)":round(est["Cobertura"],2),

        "Media":round(est["Media"],3),

        "Mediana":round(est["Mediana"],3),

        "DesvStd":round(est["Desviacion"],3),

        "CoefVar (%)":round(est["CoeficienteVariacion"],2),

        "Asimetria":round(est["Asimetria"],3),

        "Curtosis":round(est["Curtosis"],3),

        "Outliers":est["Outliers"],

        "%Outliers":round(est["PorcentajeOutliers"],2)

    })


INICIANDO ANÁLISIS UNIVARIADO

----------------------------------------------------------------------
Precipitacion_mm
----------------------------------------------------------------------
Ficha generada -> Precipitacion_mm.png

----------------------------------------------------------------------
Temp_Max_C
----------------------------------------------------------------------
Ficha generada -> Temp_Max_C.png

----------------------------------------------------------------------
Temp_Min_C
----------------------------------------------------------------------
Ficha generada -> Temp_Min_C.png

----------------------------------------------------------------------
Temp_Media_C
----------------------------------------------------------------------
Ficha generada -> Temp_Media_C.png

----------------------------------------------------------------------
Humedad_Relativa
----------------------------------------------------------------------
Ficha generada -> Humedad_Relativa.png

-----

In [40]:
#==========================================================
# DATAFRAME RESUMEN
#==========================================================

resumen = pd.DataFrame(resultado_univariado)

print()

print(resumen.head())


           Variable Origen     Categoria Unidad  Observaciones  Cobertura (%)  \
0  Precipitacion_mm  IDEAM  Climatología     mm            528          100.0   
1        Temp_Max_C  IDEAM  Climatología     °C            528          100.0   
2        Temp_Min_C  IDEAM  Climatología     °C            528          100.0   
3      Temp_Media_C  IDEAM  Climatología     °C            528          100.0   
4  Humedad_Relativa  IDEAM  Climatología      %            528          100.0   

    Media  Mediana  DesvStd  CoefVar (%)  Asimetria  Curtosis  Outliers  \
0   6.066    5.845    3.708        61.13      0.657     0.220         7   
1  22.991   22.805    2.690        11.70      0.252    -0.901         0   
2  11.103   11.260    2.475        22.29     -0.216    -0.850         0   
3  16.715   17.280    2.268        13.57     -0.162    -1.336         0   
4  84.682   86.200    5.182         6.12     -1.056     0.723         6   

   %Outliers  
0       1.33  
1       0.00  
2       0.00  
3

In [41]:
from pathlib import Path

#==========================================================
# EXPORTAR RESULTADOS
#==========================================================

CARPETA = Path("Resultados_EDA")
carpeta_resultados = CARPETA / "Fichas"

resumen.to_csv(

    carpeta_resultados /
    "estadisticas_univariadas.csv",

    index=False,

    encoding="utf-8-sig"

)

resumen.to_excel(

    carpeta_resultados /
    "estadisticas_univariadas.xlsx",

    index=False

)

resumen.to_markdown(

    carpeta_resultados /
    "estadisticas_univariadas.md",

    index=False

)

In [42]:
#==========================================================
# ÍNDICE DE FICHAS
#==========================================================

indice = resumen[

    [

        "Variable",

        "Origen",

        "Categoria",

        "Unidad",

        "Cobertura (%)"

    ]

].copy()

indice["Archivo"] = indice["Variable"].apply(

    lambda x:

    x.replace("/","-")
     .replace("\\","-")
     .replace(" ","_")
     .replace("(","")
     .replace(")","") + ".png"

)

In [43]:
indice.to_csv(

    carpeta_resultados /
    "indice_fichas.csv",

    index=False,

    encoding="utf-8-sig"

)

indice.to_excel(

    carpeta_resultados /
    "indice_fichas.xlsx",

    index=False

)

In [44]:
#==========================================================
# RESUMEN FINAL
#==========================================================

print()

print("="*80)
print("SCRIPT 16 FINALIZADO")
print("="*80)

print()

print("Variables analizadas :", len(variables))

print("Fichas generadas     :", len(variables))

print()

print("Archivos generados:")

print()

print("   Analisis_Univariado/")

print("      estadisticas_univariadas.csv")

print("      estadisticas_univariadas.xlsx")

print("      estadisticas_univariadas.md")

print("      indice_fichas.csv")

print("      indice_fichas.xlsx")

print("      *.png")


SCRIPT 16 FINALIZADO

Variables analizadas : 35
Fichas generadas     : 35

Archivos generados:

   Analisis_Univariado/
      estadisticas_univariadas.csv
      estadisticas_univariadas.xlsx
      estadisticas_univariadas.md
      indice_fichas.csv
      indice_fichas.xlsx
      *.png


In [45]:
import pandas as pd

#==========================================================
# CARGAR ESTADÍSTICAS
#==========================================================

estadisticas = pd.read_csv(
    "Resultados_EDA/Fichas/estadisticas_univariadas.csv"
)

print("Estadísticas cargadas correctamente.")

Estadísticas cargadas correctamente.


In [46]:
#==========================================================
# INDICADORES GENERALES
#==========================================================

total_variables = len(estadisticas)

cobertura_promedio = estadisticas["Cobertura (%)"].mean()

variables_baja = estadisticas[
    estadisticas["Cobertura (%)"] < 20
]

variables_alta_variabilidad = estadisticas[
    estadisticas["CoefVar (%)"] > 50
]

variables_outliers = estadisticas.sort_values(
    "%Outliers",
    ascending=False
).head(10)

variables_asimetricas = estadisticas[
    estadisticas["Asimetria"].abs() > 1
]

variables_simetricas = estadisticas[
    estadisticas["Asimetria"].abs() < 0.5
]

In [47]:
#==========================================================
# CREAR REPORTE
#==========================================================

texto = f"""
# RESUMEN EJECUTIVO DEL ANÁLISIS EXPLORATORIO DE DATOS

## Dataset

Dataset Maestro V04

## Variables analizadas

{total_variables}

## Cobertura promedio

{cobertura_promedio:.2f} %

## Variables con cobertura menor al 20 %

{len(variables_baja)}

{variables_baja["Variable"].to_string(index=False)}

## Variables con alta variabilidad

{len(variables_alta_variabilidad)}

{variables_alta_variabilidad["Variable"].to_string(index=False)}

## Variables con mayor presencia de valores atípicos

{variables_outliers[["Variable","%Outliers"]].to_markdown(index=False)}

## Variables aproximadamente simétricas

{variables_simetricas["Variable"].to_string(index=False)}

## Variables altamente asimétricas

{variables_asimetricas["Variable"].to_string(index=False)}

## Conclusiones

El Dataset Maestro V04 presenta una estructura consistente
para continuar hacia el análisis bivariado y el modelado.

Las variables con baja cobertura fueron identificadas y
documentadas para tratamientos específicos.

Las variables con mayor variabilidad y presencia de valores
atípicos deberán ser consideradas durante la construcción
de modelos predictivos.

El conjunto de datos se considera apto para continuar con
la siguiente etapa del Framework correspondiente al
Análisis Bivariado y Correlaciones.
"""

In [48]:
#==========================================================
# EXPORTAR
#==========================================================

with open(

    "Resultados_EDA/Resumen_Ejecutivo_EDA.md",

    "w",

    encoding="utf-8"

) as archivo:

    archivo.write(texto)

print()

print("Resumen Ejecutivo generado correctamente.")


Resumen Ejecutivo generado correctamente.


## **M6. Análisis Bivariado y Selección de Variables del Dataset Maestro V04**

In [49]:
#==========================================================
# FRAMEWORK PARA LA GESTIÓN INTELIGENTE DEL RECURSO HÍDRICO
#
#
# Objetivo:
#     Analizar las relaciones entre las variables
#     numéricas del Dataset Maestro V04 mediante
#     correlaciones de Pearson y Spearman,
#     identificar variables altamente correlacionadas
#     y generar automáticamente reportes y
#     visualizaciones para apoyar la selección
#     de variables del Framework.
#
# Autor:
#     Juan Carlos Riatiga Peña
#
# Maestría en Ciencia de Datos
#==========================================================

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

from pathlib import Path

url = (
    "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"
)

print("Cargando Dataset Maestro V04...")

dataset = pd.read_parquet(url)

print()

print("Dataset cargado correctamente.")

print()

print(dataset.shape)

Cargando Dataset Maestro V04...

Dataset cargado correctamente.

(528, 42)


In [51]:
#==========================================================
# CARPETA DE RESULTADOS
#==========================================================

carpeta = Path("Resultados_EDA/Analisis_Bivariado")

carpeta.mkdir(
    exist_ok=True
)

print()

print("Carpeta de resultados creada.")


Carpeta de resultados creada.


In [52]:
#==========================================================
# VARIABLES NUMÉRICAS
#==========================================================

variables = dataset.select_dtypes(
    include=np.number
).columns.tolist()

print()

print("Variables numéricas encontradas:", len(variables))

for variable in variables:
    print("-", variable)


Variables numéricas encontradas: 35
- Precipitacion_mm
- Temp_Max_C
- Temp_Min_C
- Temp_Media_C
- Humedad_Relativa
- Velocidad_Viento
- Radiacion_Solar
- Anio
- Mes
- ONI
- VolumenUtilDiarioMasa
- irca
- POBLACION TOTAL
- DENSIDAD POBLACIONAL
- INDICE DE DESEMPENO INSTITUCIONAL
- ACCESO A AGUA POTABLE ADECUADO
- PORCENTAJE DE LA POBLACION CON ACCESO A METODOS DE SANEAMIENTO ADECUADOS
- COBERTURA DE ACUEDUCTO URBANO
- COBERTURA DE ACUEDUCTO RURAL
- COBERTURA DE ALCANTARILLADO URBANO
- COBERTURA DE ALCANTARILLADO RURAL
- CONTINUIDAD DE ACUEDUCTO URBANO
- CALIDAD DEL AGUA
- AGUAS RESIDUALES TRATADAS
- Nivel_Minimo
- CONDUCTIVIDAD ELECTRICA
- DEMANDA BIOQUIMICA DE OXIGENO (DBO5)
- DEMANDA QUIMICA DE OXIGENO (DQO)
- FOSFORO TOTAL
- NITROGENO TOTAL
- OXIGENO DISUELTO (OD)
- SOLIDOS SUSPENDIDOS TOTALES
- TEMPERATURA
- TURBIDEZ
- pH


In [53]:
#==========================================================
# MATRICES DE CORRELACIÓN
#==========================================================

print()

print("Calculando correlaciones...")

correlacion_pearson = dataset[
    variables
].corr(
    method="pearson"
)

correlacion_spearman = dataset[
    variables
].corr(
    method="spearman"
)

print()

print("Matrices calculadas correctamente.")


Calculando correlaciones...

Matrices calculadas correctamente.


In [54]:
#==========================================================
# EXPORTAR MATRICES
#==========================================================

correlacion_pearson.to_csv(
    carpeta/"correlacion_pearson.csv",
    encoding="utf-8-sig"
)

correlacion_pearson.to_excel(
    carpeta/"correlacion_pearson.xlsx"
)

correlacion_pearson.to_markdown(
    carpeta/"correlacion_pearson.md"
)

correlacion_spearman.to_csv(
    carpeta/"correlacion_spearman.csv",
    encoding="utf-8-sig"
)

correlacion_spearman.to_excel(
    carpeta/"correlacion_spearman.xlsx"
)

correlacion_spearman.to_markdown(
    carpeta/"correlacion_spearman.md"
)

print()

print("Matrices exportadas.")


Matrices exportadas.


In [55]:
#==========================================================
# HEATMAP PEARSON
#==========================================================

plt.figure(figsize=(18,15))

plt.imshow(
    correlacion_pearson,
    interpolation="nearest",
    aspect="auto"
)

plt.colorbar()

plt.xticks(
    range(len(variables)),
    variables,
    rotation=90,
    fontsize=8
)

plt.yticks(
    range(len(variables)),
    variables,
    fontsize=8
)

plt.title(
    "Matriz de Correlación Pearson"
)

plt.tight_layout()

plt.savefig(
    carpeta/"heatmap_pearson.png",
    dpi=300
)

plt.close()

print("Heatmap Pearson generado.")

Heatmap Pearson generado.


In [56]:
#==========================================================
# HEATMAP SPEARMAN
#==========================================================

plt.figure(figsize=(18,15))

plt.imshow(
    correlacion_spearman,
    interpolation="nearest",
    aspect="auto"
)

plt.colorbar()

plt.xticks(
    range(len(variables)),
    variables,
    rotation=90,
    fontsize=8
)

plt.yticks(
    range(len(variables)),
    variables,
    fontsize=8
)

plt.title(
    "Matriz de Correlación Spearman"
)

plt.tight_layout()

plt.savefig(
    carpeta/"heatmap_spearman.png",
    dpi=300
)

plt.close()

print("Heatmap Spearman generado.")

Heatmap Spearman generado.


In [57]:
#==========================================================
# RANKING DE CORRELACIONES
#==========================================================

ranking = []

for i in range(len(variables)):

    for j in range(i+1,len(variables)):

        ranking.append([

            variables[i],

            variables[j],

            correlacion_pearson.iloc[i,j],

            correlacion_spearman.iloc[i,j]

        ])

ranking = pd.DataFrame(

    ranking,

    columns=[

        "Variable_1",

        "Variable_2",

        "Pearson",

        "Spearman"

    ]

)

In [58]:
ranking["Abs_Pearson"] = ranking["Pearson"].abs()

ranking["Abs_Spearman"] = ranking["Spearman"].abs()

In [59]:
ranking = ranking.sort_values(

    "Abs_Pearson",

    ascending=False

)

In [60]:
#==========================================================
# CLASIFICAR CORRELACIÓN
#==========================================================

def clasificar(valor):

    valor = abs(valor)

    if valor >= 0.90:
        return "Muy fuerte"

    elif valor >= 0.70:
        return "Fuerte"

    elif valor >= 0.50:
        return "Moderada"

    elif valor >= 0.30:
        return "Débil"

    else:
        return "Muy débil"

ranking["Clasificacion"] = ranking["Pearson"].apply(clasificar)

In [61]:
ranking.to_csv(

    carpeta/"ranking_correlaciones.csv",

    index=False,

    encoding="utf-8-sig"

)

ranking.to_excel(

    carpeta/"ranking_correlaciones.xlsx",

    index=False

)

ranking.to_markdown(

    carpeta/"ranking_correlaciones.md",

    index=False

)

In [62]:
#==========================================================
# VARIABLES ALTAMENTE CORRELACIONADAS
#==========================================================

umbral = 0.90

variables_redundantes = ranking[
    ranking["Abs_Pearson"] >= umbral
].copy()

print()

print("Variables altamente correlacionadas:")

print(len(variables_redundantes))


Variables altamente correlacionadas:
32


In [63]:
#==========================================================
# RECOMENDACIÓN
#==========================================================

def recomendar(r):

    if abs(r) >= 0.95:

        return "Considerar conservar solo una variable."

    elif abs(r) >= 0.90:

        return "Revisar posible redundancia."

    else:

        return "Conservar ambas."

variables_redundantes["Recomendacion"] = (

    variables_redundantes["Pearson"]

    .apply(recomendar)

)

In [64]:
variables_redundantes.to_csv(

    carpeta/"variables_redundantes.csv",

    index=False,

    encoding="utf-8-sig"

)

variables_redundantes.to_excel(

    carpeta/"variables_redundantes.xlsx",

    index=False

)

variables_redundantes.to_markdown(

    carpeta/"variables_redundantes.md",

    index=False

)

In [65]:
#==========================================================
# INFORME AUTOMÁTICO
#==========================================================

texto = []

texto.append("# Informe de Correlaciones\n")

texto.append(f"Variables analizadas: {len(variables)}")

texto.append(
    f"\nPares evaluados: {len(ranking)}"
)

texto.append(
    f"\nVariables altamente correlacionadas: {len(variables_redundantes)}"
)

texto.append(
    "\nSe recomienda revisar estas variables antes de la etapa de Machine Learning."
)

In [66]:
texto.append("\n\n## Top 20 correlaciones\n")

top20 = ranking.head(20)

for _, fila in top20.iterrows():

    texto.append(

        f"- {fila['Variable_1']}  <->  "
        f"{fila['Variable_2']}  "
        f"(Pearson={fila['Pearson']:.3f})"

    )

In [67]:
with open(

    carpeta/"Informe_Correlaciones.md",

    "w",

    encoding="utf-8"

) as archivo:

    archivo.write("\n".join(texto))

In [68]:
indice = pd.DataFrame({

    "Archivo":[

        "correlacion_pearson.csv",

        "correlacion_spearman.csv",

        "heatmap_pearson.png",

        "heatmap_spearman.png",

        "ranking_correlaciones.csv",

        "variables_redundantes.csv",

        "Informe_Correlaciones.md"

    ],

    "Descripción":[

        "Matriz Pearson",

        "Matriz Spearman",

        "Mapa de calor Pearson",

        "Mapa de calor Spearman",

        "Ranking de correlaciones",

        "Variables redundantes",

        "Informe automático"

    ]

})

indice.to_csv(

    carpeta/"indice_resultados.csv",

    index=False,

    encoding="utf-8-sig"

)

indice.to_excel(

    carpeta/"indice_resultados.xlsx",

    index=False

)

In [69]:
print()

print("="*80)
print("FINALIZADO")
print("="*80)

print()

print("Variables analizadas :", len(variables))
print("Pares evaluados      :", len(ranking))
print("Redundancias         :", len(variables_redundantes))

print()

print("Archivos generados:")

print()

print("   Analisis_Bivariado/")
print("      correlacion_pearson.csv")
print("      correlacion_spearman.csv")
print("      heatmap_pearson.png")
print("      heatmap_spearman.png")
print("      ranking_correlaciones.csv")
print("      variables_redundantes.csv")
print("      Informe_Correlaciones.md")
print("      indice_resultados.csv")


FINALIZADO

Variables analizadas : 35
Pares evaluados      : 595
Redundancias         : 32

Archivos generados:

   Analisis_Bivariado/
      correlacion_pearson.csv
      correlacion_spearman.csv
      heatmap_pearson.png
      heatmap_spearman.png
      ranking_correlaciones.csv
      variables_redundantes.csv
      Informe_Correlaciones.md
      indice_resultados.csv


**Análisis Integrado del Comportamiento del Dataset Maestro V04**

**Calidad general del conjunto de datos**

El Dataset Maestro V04 quedó conformado por **35 variables** provenientes de múltiples fuentes oficiales nacionales (IDEAM, DANE, SUI, DNP, INS, ENSO y registros institucionales), integradas mediante el Framework desarrollado.

La cobertura promedio alcanzó aproximadamente el **61.8%**, lo cual es adecuado considerando la heterogeneidad temporal y espacial de las fuentes.

> ⚠️ **Decisión Metodológica Clave:** Durante la etapa de ingeniería de datos se decidió no imputar aquellas variables cuya disponibilidad era inferior al 20%, evitando introducir información artificial que pudiera afectar el comportamiento real del sistema. Esta decisión fortalece la confiabilidad científica del conjunto de datos.

---

**Comportamiento univariado**

El análisis univariado permitió identificar cuatro grandes características del Dataset Maestro:

*   **a) Variables de alta cobertura:** Las variables climáticas e hidrológicas presentan prácticamente cobertura completa. Entre ellas: `Precipitación`, `Temperaturas`, `Humedad`, `Radiación Solar`, `ENSO`, `Volumen útil` y `Variables demográficas`. Estas variables constituyen el núcleo del modelo predictivo.
*   **b) Variables de cobertura limitada:** Se identificaron quince variables con coberturas inferiores al 20%. Entre ellas: `DBO5`, `DQO`, `Fósforo`, `Nitrógeno`, `Turbidez`, `Conductividad`, `Oxígeno disuelto`, `pH` e `Índice de Desempeño Institucional`. Estas variables fueron conservadas por su importancia ambiental, aunque serán utilizadas únicamente en modelos donde exista suficiente información disponible.
*   **c) Variabilidad:** Las variables relacionadas con calidad del agua presentan elevados coeficientes de variación. Esto indica que los cuerpos de agua evaluados presentan condiciones muy diferentes entre municipios y periodos. No representa un error; por el contrario, refleja la heterogeneidad ambiental propia del sistema hídrico colombiano.
*   **d) Valores atípicos:** El análisis mediante IQR permitió detectar valores extremos principalmente en `DBO5`, `Turbidez`, `Conductividad`, `Calidad del agua` e `IRCA`. En variables ambientales estos valores suelen corresponder a eventos reales (temporadas de lluvia, sequías, vertimientos, contaminación localizada), por lo que **no se recomienda eliminarlos automáticamente**.

---

**Forma de las distribuciones**

Las fichas estadísticas mostraron que el Dataset no sigue un comportamiento normal. Se observan tres grupos principales:

*   **Variables aproximadamente simétricas:** Principalmente variables climáticas (v.g., `Temperatura`, `Radiación`, `Año`, `Mes`). Estas variables son adecuadas para modelos lineales tradicionales.
*   **Variables altamente asimétricas:** Principalmente `IRCA`, `DBO5`, `Turbidez`, `Conductividad`, `Nitrógeno` y `Fósforo`. Estas distribuciones son esperables debido a que representan fenómenos ambientales con ocurrencia esporádica y eventos extremos.
*   **Variables con colas largas:** Las variables de calidad del agua presentan curtosis elevada. Esto confirma la presencia de observaciones extremas asociadas a eventos poco frecuentes.

---

**Relaciones bivariadas**

El análisis de correlaciones evaluó **595 pares de variables**, de los cuales se identificaron **32 pares con correlaciones muy fuertes**. Los resultados muestran tres comportamientos claramente diferenciados:

*   **Grupo 1 | Variables de cobertura de servicios públicos:** Se identificaron correlaciones prácticamente perfectas entre `Densidad poblacional`, `Cobertura de acueducto`, `Cobertura de alcantarillado`, `Acceso a agua potable` y `Acceso a saneamiento`, alcanzando valores cercanos a 1.0. Esto indica una fuerte dependencia entre estos indicadores. No significa un error, ya que todas provienen del mismo contexto demográfico e institucional. Sin embargo, para Machine Learning existe un riesgo latente de **multicolinealidad**, por lo que será necesario seleccionar únicamente las variables más representativas.
*   **Grupo 2 | Variables de calidad del agua:** Se observaron asociaciones muy fuertes entre `Conductividad`, `Nitrógeno`, `Fósforo` y otras variables fisicoquímicas. Estas relaciones son consistentes con procesos de contaminación del recurso hídrico y son completamente coherentes desde el punto de vista ambiental.
*   **Grupo 3 | Variables hidrológicas:** Se encontró una correlación elevada entre `Volumen útil diario` e `IRCA` (Pearson $\approx 0.95$). Este resultado sugiere que la disponibilidad del recurso hídrico podría estar asociada con la calidad del agua, aspecto que será evaluado posteriormente mediante modelos predictivos.

---

**Correlaciones perfectas**

Uno de los hallazgos más relevantes corresponde a la aparición de correlaciones exactamente iguales a 1. En un análisis superficial podría interpretarse como un error; sin embargo, en este proyecto existen varias razones metodológicas que lo explican:
1. Variables derivadas de una misma fuente oficial.
2. Indicadores calculados a partir de una misma población base.
3. Expansión temporal de indicadores anuales a registros mensuales.
4. Variables que representan distintas formas de medir un mismo fenómeno.

> **Conclusión:** Estas correlaciones deben interpretarse como evidencia de redundancia informativa y no necesariamente como errores del proceso de integración.

---

**Implicaciones para el modelado**

A partir del análisis exploratorio pueden establecerse las siguientes decisiones metodológicas:

*   ✔️ Conservar todas las variables durante la fase exploratoria.
*   ✔️ Evaluar la reducción de dimensionalidad antes del entrenamiento de modelos.
*   ✔️ Controlar la multicolinealidad mediante análisis de **VIF** (Factor de Inflación de la Varianza) o importancia de variables.
*   ✔️ Mantener los valores atípicos por corresponder a fenómenos ambientales reales.
*   ✔️ Emplear algoritmos robustos a distribuciones no normales (por ejemplo, **Random Forest, XGBoost o LightGBM**), evitando asumir normalidad cuando no existe.

---

**Conclusión general**

El análisis integrado evidencia que el **Dataset Maestro V04** posee una estructura consistente y científicamente válida para continuar hacia la fase de preparación para Machine Learning. Las principales limitaciones identificadas corresponden a la baja cobertura de algunas variables ambientales y a la presencia de grupos de variables altamente correlacionadas, condiciones esperables dada la naturaleza de las fuentes de información y del dominio de estudio.

Estas características no comprometen la calidad del conjunto de datos; por el contrario, orientan las decisiones metodológicas que deberán adoptarse en la etapa de modelado para evitar problemas de multicolinealidad, preservar la información relevante y maximizar la capacidad predictiva del framework propuesto.
